File: apps/mobile/targets/positions-widget/Sources/Rendering/Common/Stats/StatBlock.swift
import SwiftUI

/// Vertical stat group with a header label and two stacked value lines
/// (primary above secondary). Used by the Medium widget's right column
/// for `Value` and `Fees` sections.
struct StatBlock: View {
  let label: String
  let accent: Color
  let primary: TokenAmount
  let secondary: TokenAmount
  let density: Density

  enum Density {
    /// Display density — used for the headline `Value` block.
    case display
    /// Compact density — used for the secondary `Fees` block.
    case compact

    var primaryFont: Font {
      switch self {
      case .display: return TypeScale.valueLg
      case .compact: return TypeScale.valueSm
      }
    }

    var secondaryFont: Font {
      switch self {
      case .display: return TypeScale.valueXs
      case .compact: return TypeScale.valueXxxs
      }
    }

    var primarySymbolFont: Font {
      switch self {
      case .display: return TypeScale.suffixLg
      case .compact: return TypeScale.suffixSm
      }
    }

    var secondarySymbolFont: Font {
      switch self {
      case .display: return TypeScale.suffixMd
      case .compact: return TypeScale.suffixSm
      }
    }
  }

  var body: some View {
    VStack(alignment: .leading, spacing: Spacing.xxs) {
      Text(label)
        .font(TypeScale.labelMd)
        .foregroundStyle(Color.textMuted)
      StatLine(
        amount: primary,
        valueFont: density.primaryFont,
        symbolFont: density.primarySymbolFont,
        valueColor: accent
      )
      StatLine(
        amount: secondary,
        valueFont: density.secondaryFont,
        symbolFont: density.secondarySymbolFont,
        valueColor: accent
      )
    }
  }
}

File: apps/mobile/modules/widget-bridge/android/src/main/java/expo/modules/widgetbridge/widget/ui/components/StatBlock.kt
package expo.modules.widgetbridge.widget.ui.components

import androidx.compose.runtime.Composable
import androidx.compose.ui.graphics.Color
import androidx.glance.layout.Column
import androidx.glance.text.Text
import androidx.glance.text.TextStyle
import expo.modules.widgetbridge.widget.theme.Colors
import expo.modules.widgetbridge.widget.theme.Typography

enum class Density(
  val primary: TextStyle,
  val secondary: TextStyle,
  val primarySymbol: TextStyle,
  val secondarySymbol: TextStyle,
) {
  Display(Typography.valueLg, Typography.valueXs, Typography.suffixLg, Typography.suffixMd),
  Compact(Typography.valueSm, Typography.valueXxxs, Typography.suffixSm, Typography.suffixSm),
}

@Composable
fun StatBlock(
  label: String,
  accent: Color,
  primary: TokenAmount,
  secondary: TokenAmount,
  density: Density,
) {
  Column {
    Text(
      text = label,
      style = Typography.withColor(Typography.labelMd, Colors.textMuted),
      maxLines = 1,
    )
    StatLine(primary, density.primary, density.primarySymbol, accent)
    StatLine(secondary, density.secondary, density.secondarySymbol, accent)
  }
}
